src/query/expression/src/function.rs

```diff
@@ -153,29 +153,73 @@ pub struct FunctionRegistry {
 }
 
 impl Function {
-    pub fn wrap_nullable(self) -> Self {
-        let (_, eval) = self.eval.into_scalar().unwrap();
-        Self {
-            signature: FunctionSignature {
-                name: self.signature.name.clone(),
-                args_type: self
-                    .signature
-                    .args_type
-                    .iter()
-                    .map(|ty| ty.wrap_nullable())
-                    .collect(),
-                return_type: self.signature.return_type.wrap_nullable(),
-            },
+    pub fn passthrough_nullable(self) -> Self {
+        debug_assert!(
+            !self
+                .signature
+                .args_type
+                .iter()
+                .any(|ty| ty.is_nullable_or_null())
+        );
+
+        let (calc_domain, eval) = self.eval.into_scalar().unwrap();
+
+        let signature = FunctionSignature {
+            name: self.signature.name.clone(),
+            args_type: self
+                .signature
+                .args_type
+                .iter()
+                .map(|ty| ty.wrap_nullable())
+                .collect(),
+            return_type: self.signature.return_type.wrap_nullable(),
+        };
+
+        let new_calc_domain = Box::new(move |ctx: &FunctionContext, domains: &[Domain]| {
+            if domains
+                .iter()
+                .any(|domain| domain.as_nullable().unwrap().value.is_none())
+            {
+                return FunctionDomain::Domain(Domain::Nullable(NullableDomain {
+                    has_null: true,
+                    value: None,
+                }));
+            }
+
+            let inner_domains = domains
+                .iter()
+                .map(|domain| (**domain.as_nullable().unwrap().value.as_ref().unwrap()).clone())
+                .collect::<Vec<_>>();
+            let result = calc_domain(ctx, &inner_domains);
+            result.map(|domain| {
+                let (result_has_null, result_domain) = match domain {
+                    Domain::Nullable(NullableDomain { has_null, value }) => (has_null, value),
+                    domain => (false, Some(Box::new(domain))),
+                };
+                let has_null = result_has_null
+                    || domains
+                        .iter()
+                        .any(|domain| domain.as_nullable().unwrap().has_null);
```

Comment:
I think it's better to merge the three iterator operations into one. (L179~L181, L189~L192, L200~L202)

In another word, collect the necessary variables in one for loop.